From a2ed16da892be89f54d944a6e24c3a155419ac07 Mon Sep 17 00:00:00 2001
From: "sun.cwsun" <sun.cwsun@inventec.com>
Date: Fri, 20 May 2022 12:02:11 +0800
Subject: [PATCH 3/6] Bug 1016 [SW][Transformers][OpenBMC] MS M-BMC-LOG0008 The
 BMC should generate an BMC Health SEL when firmware upgrade

		-  The BMC should generate an BMC Health SEL when firmware upgrade

	Root Cause:
        - none

    Solution/Change:
        [phosphor-software-manager]
			- do MS SEL log when start flashing firmware and flashing completed

    Entry Test:
		- BMC firmware upgrading  via OpenBMC webui
		-- flashing starting SEL log
sysadmin@transformers:~# ipmitool sel get 0x20d
SEL Record ID          : 020d
 Record Type           : 02
 Timestamp             : 13:54:32 GMT 13:54:32 GMT
 Generator ID          : 0020
 EvM Revision          : 04
 Sensor Type           : Unknown
 Sensor Number         : b1
 Event Type            : OEM
 Event Direction       : Assertion Event
 Event Data (RAW)      : ae0100
 Description           :

Sensor ID              : BMC_health (0xb1)
 Entity ID             : 7.0
 Sensor Type (Discrete): Unknown

		-- flashing completed SEL log
		sysadmin@transformers:~# ipmitool sel get 0x28f
SEL Record ID          : 028f
 Record Type           : 02
 Timestamp             : 14:19:00 GMT 14:19:00 GMT
 Generator ID          : 0020
 EvM Revision          : 04
 Sensor Type           : Unknown
 Sensor Number         : b1
 Event Type            : OEM
 Event Direction       : Assertion Event
 Event Data (RAW)      : af0100
 Description           :

Sensor ID              : BMC_health (0xb1)
 Entity ID             : 7.0
 Sensor Type (Discrete): Unknown

	- BIOS firmware upgrade by curl
		-- starting bios firmware upgrade

curl -k -basic -u root:0penBmc -X POST \
https://10.6.141.193/redfish/v1/UpdateService/\
Actions/UpdateService.SimpleUpdate -d \
'{"ImageURI": "tftp://10.6.142.48/0ACOR041_563_R_20211229_AST2600.tar.gz"}'

		-- get sel result
		sysadmin@transformers:~# ipmitool sel get 0x2ec
SEL Record ID          : 02ec
 Record Type           : 02
 Timestamp             : 03:21:21 GMT 03:21:21 GMT
 Generator ID          : 0020
 EvM Revision          : 04
 Sensor Type           : Unknown
 Sensor Number         : b1
 Event Type            : OEM
 Event Direction       : Assertion Event
 Event Data (RAW)      : ae0200
 Description           :

Sensor ID              : BMC_health (0xb1)
 Entity ID             : 7.0
 Sensor Type (Discrete): Unknown

	sysadmin@transformers:~# ipmitool sel get 0x2ed
SEL Record ID          : 02ed
 Record Type           : 02
 Timestamp             : 03:21:21 GMT 03:21:21 GMT
 Generator ID          : 0020
 EvM Revision          : 04
 Sensor Type           : Unknown
 Sensor Number         : b1
 Event Type            : OEM
 Event Direction       : Assertion Event
 Event Data (RAW)      : af0200
 Description           :

Sensor ID              : BMC_health (0xb1)
 Entity ID             : 7.0
 Sensor Type (Discrete): Unknown

	- upgrade CPLD by redfish command
		-- cwsun@myubuntu:~$ curl -k -basic -u root:0penBmc -X POST https://10.6.141.193/redfish/v1/UpdateService/Actions/UpdateService.SimpleUpdate -d '{"ImageURI": "tftp://10.6.142.48/Transformer_SCM_cpld_dvt_20211221_a_usercode0x24_FormalRelease.tar.gz"}'
{
  "@odata.id": "/redfish/v1/TaskService/Tasks/1",
  "@odata.type": "#Task.v1_4_3.Task",
  "Id": "1",
  "TaskState": "Running",
  "TaskStatus": "OK"
}cwsun@myubuntu:~$

	sysadmin@transformers:~# ipmitool sel get 0x308
SEL Record ID          : 0308
 Record Type           : 02
 Timestamp             : 03:55:18 GMT 03:55:18 GMT
 Generator ID          : 0020
 EvM Revision          : 04
 Sensor Type           : Unknown
 Sensor Number         : b1
 Event Type            : OEM
 Event Direction       : Assertion Event
 Event Data (RAW)      : ae0300
 Description           :

Sensor ID              : BMC_health (0xb1)
 Entity ID             : 7.0
 Sensor Type (Discrete): Unknown

sysadmin@transformers:~# ipmitool sel get 0x309
SEL Record ID          : 0309
 Record Type           : 02
 Timestamp             : 03:55:18 GMT 03:55:18 GMT
 Generator ID          : 0020
 EvM Revision          : 04
 Sensor Type           : Unknown
 Sensor Number         : b1
 Event Type            : OEM
 Event Direction       : Assertion Event
 Event Data (RAW)      : af0300
 Description           :

Sensor ID              : BMC_health (0xb1)
 Entity ID             : 7.0
 Sensor Type (Discrete): Unknown

---
 activation.cpp        | 12 ++++++++++++
 item_updater_main.cpp | 17 +++++++++++++++++
 mmc/flash.cpp         |  9 +++++++++
 static/flash.cpp      |  8 ++++++++
 ubi/flash.cpp         |  8 ++++++++
 utils.cpp             | 20 ++++++++++++++++++++
 utils.hpp             |  8 ++++++++
 7 files changed, 82 insertions(+)

diff --git a/activation.cpp b/activation.cpp
index e434f3b..7679b4d 100644
--- a/activation.cpp
+++ b/activation.cpp
@@ -422,6 +422,9 @@ bool Activation::checkApplyTimeImmediate()
 #ifdef HOST_BIOS_UPGRADE
 void Activation::flashWriteHost()
 {
+    std::vector<uint8_t> eventData = {0xAE, 0x02, 0x00};
+    utils::reportBySEL(this->bus, eventData);
+
     std::string biosImagePath;
     auto ret = getImagePath(biosImagePath);
     if(ret!=0)
@@ -453,6 +456,9 @@ void Activation::flashWriteHost()
 
 void Activation::onStateChangesBios(sdbusplus::message_t& msg)
 {
+    std::vector<uint8_t> eventData = {0xAF, 0x02, 0x00};
+    utils::reportBySEL(this->bus, eventData);
+
     uint32_t newStateID{};
     sdbusplus::message::object_path newStateObjPath;
     std::string newStateUnit{};
@@ -525,6 +531,9 @@ int32_t Activation::getImagePath(std::string& imagePath)
 
 void Activation::flashWriteCpld()
 {
+    std::vector<uint8_t> eventData = {0xAE, 0x03, 0x00};
+    utils::reportBySEL(this->bus, eventData);
+
     std::string cpldImagePath;
     auto ret = getImagePath(cpldImagePath);
     if(ret!=0)
@@ -556,6 +565,9 @@ void Activation::flashWriteCpld()
 
 void Activation::onStateChangesCpld(sdbusplus::message::message& msg)
 {
+    std::vector<uint8_t> eventData = {0xAF, 0x03, 0x00};
+    utils::reportBySEL(this->bus, eventData);
+
     uint32_t newStateID{};
     sdbusplus::message::object_path newStateObjPath;
     std::string newStateUnit{};
diff --git a/item_updater_main.cpp b/item_updater_main.cpp
index 790af60..445253c 100644
--- a/item_updater_main.cpp
+++ b/item_updater_main.cpp
@@ -7,6 +7,9 @@
 #include <sdbusplus/bus.hpp>
 #include <sdbusplus/server/manager.hpp>
 
+#include <iostream>
+#include <fstream>
+
 boost::asio::io_context& getIOContext()
 {
     static boost::asio::io_context io;
@@ -17,6 +20,20 @@ int main()
 {
     sdbusplus::asio::connection bus(getIOContext());
 
+    //if the /var/lib/os-release overwrite_existing
+    //The firmware is update completed
+    FILE *fp = fopen("/var/lib/os-release", "r");
+
+    if( fp != NULL)
+    {
+        //do MS SEL firmware upgrade complete and remove the file
+        fclose(fp);
+        remove("/var/lib/os-release");
+
+        std::vector<uint8_t> eventData = {0xAF, 0x01, 0x00};
+        utils::reportBySEL(bus, eventData);
+    }
+
     // Add sdbusplus ObjectManager.
     sdbusplus::server::manager_t objManager(bus, SOFTWARE_OBJPATH);
 
diff --git a/mmc/flash.cpp b/mmc/flash.cpp
index 3bc2ab8..194c617 100644
--- a/mmc/flash.cpp
+++ b/mmc/flash.cpp
@@ -5,6 +5,10 @@
 #include "activation.hpp"
 #include "item_updater.hpp"
 
+#include <filesystem>
+
+namespace fs = std::filesystem;
+
 namespace phosphor
 {
 namespace software
@@ -16,6 +20,11 @@ namespace softwareServer = sdbusplus::xyz::openbmc_project::Software::server;
 
 void Activation::flashWrite()
 {
+    auto copyOption = fs::copy_options::overwrite_existing;
+    fs::copy("/usr/lib/os-release", "/var/lib/os-release");
+    std::vector<uint8_t> eventData = {0xAE, 0x01, 0x00};
+    utils::reportBySEL(bus, eventData);
+
     auto method = bus.new_method_call(SYSTEMD_BUSNAME, SYSTEMD_PATH,
                                       SYSTEMD_INTERFACE, "StartUnit");
     auto serviceFile = "obmc-flash-mmc@" + versionId + ".service";
diff --git a/static/flash.cpp b/static/flash.cpp
index 74316d1..62336d9 100644
--- a/static/flash.cpp
+++ b/static/flash.cpp
@@ -11,6 +11,8 @@
 #include <filesystem>
 #include <system_error>
 
+namespace fs = std::filesystem;
+
 namespace
 {
 constexpr auto PATH_INITRAMFS = "/run/initramfs";
@@ -45,6 +47,12 @@ void Activation::flashWrite()
         return;
     }
 #endif
+
+    auto copyOption = fs::copy_options::overwrite_existing;
+    fs::copy("/usr/lib/os-release", "/var/lib/os-release", copyOption);
+    std::vector<uint8_t> eventData = {0xAE, 0x01, 0x00};
+    utils::reportBySEL(bus, eventData);
+
     // For static layout code update, just put images in /run/initramfs.
     // It expects user to trigger a reboot and an updater script will program
     // the image to flash during reboot.
diff --git a/ubi/flash.cpp b/ubi/flash.cpp
index b445a56..b8ddac6 100644
--- a/ubi/flash.cpp
+++ b/ubi/flash.cpp
@@ -2,6 +2,9 @@
 
 #include "activation.hpp"
 #include "item_updater.hpp"
+#include <filesystem>
+
+namespace fs = std::filesystem;
 
 namespace phosphor
 {
@@ -14,6 +17,11 @@ namespace softwareServer = sdbusplus::xyz::openbmc_project::Software::server;
 
 void Activation::flashWrite()
 {
+    auto copyOption = fs::copy_options::overwrite_existing;
+    fs::copy("/usr/lib/os-release", "/var/lib/os-release");
+    std::vector<uint8_t> eventData = {0xAE, 0x01, 0x00};
+    utils::reportBySEL(bus, eventData);
+
     auto method = bus.new_method_call(SYSTEMD_BUSNAME, SYSTEMD_PATH,
                                       SYSTEMD_INTERFACE, "StartUnit");
     method.append("obmc-flash-bmc-ubirw.service", "replace");
diff --git a/utils.cpp b/utils.cpp
index 204affa..2f338d5 100644
--- a/utils.cpp
+++ b/utils.cpp
@@ -121,4 +121,24 @@ std::pair<int, std::string> executeCmd(char** args)
 
 } // namespace internal
 
+void reportBySEL(sdbusplus::bus::bus& bus, std::vector<uint8_t>& eventData)
+{
+    sdbusplus::message::message writeSEL = bus.new_method_call(
+        "xyz.openbmc_project.Logging.IPMI", "/xyz/openbmc_project/Logging/IPMI",
+        "xyz.openbmc_project.Logging.IPMI", "IpmiSelAdd");
+
+    writeSEL.append(
+        std::string("BMC Health check"),
+        "/xyz/openbmc_project/sensors/oem_event_70h/oem_e0h/BMC_health",
+        eventData, true, static_cast<uint16_t>(0x20));
+    try
+    {
+        bus.call(writeSEL);
+    }
+    catch (sdbusplus::exception_t& e)
+    {
+        fprintf(stderr, "%s:%d exception=%s \n", __func__, __LINE__, e.what());
+    }
+}
+
 } // namespace utils
diff --git a/utils.hpp b/utils.hpp
index be4fb2c..91bfa96 100644
--- a/utils.hpp
+++ b/utils.hpp
@@ -117,4 +117,12 @@ std::pair<int, std::string> execute(const char* path, Arg&&... args)
     return internal::executeCmd(argArray.data());
 }
 
+/**
+ * @brief do system sel log for system firmware update
+ *
+ * @param bus the dbus bus object
+ * @param eventData the eventData of system sel
+ */
+void reportBySEL(sdbusplus::bus::bus& bus, std::vector<uint8_t>& eventData);
+
 } // namespace utils
-- 
2.25.1

